### ISSUE TITLE
Autocompletion for citekeys

### ISSUE THREAD

**Issue by 73** [pubs/pubs#34]
Hi Guys,

I would like suggest autocompletion for (well that's what I would like) zsh.
This might be a place to start: https://github.com/zsh-users/zsh-completions

Let me know what you think.
--73

**Comment by humm** [comment]
Yes, autocompletion would be nice. For the moment, we are looking into basic command autocompletion using https://github.com/kislyuk/argcomplete, which is zsh compatible. but citekey autocompletion is the next step.

**Comment by humm** [comment]
FYI, I looked and tried to add autocompletion with argcomplete, but it is dicier than it looks. Given the current code, argcomplete is not a good option.

**Comment by 73** [comment]
Thank you for looking l. I really appreciate that. What is the problem?

**Comment by humm** [comment]
Besides the fact that I did not actually manage to make it work, argcomplete actually execute the program up until parse_args(). I guess our double parser solution is wreaking havoc in argcomplete, but it also means that the conf needs to be loaded everytime the user hits tab, which is way too slow and produce noticeable delays in the terminal.

**Comment by 73** [comment]
Ah okay. If the commands are stable we could write a auto complete function
shell only. Or is there another option?

**Comment by humm** [comment]
IDK

**Comment by omangin** [comment]
An issue will be that with a lot of papers it is very long to parse all files. This will be solved by the cache but with the current implementation `list` reads all files even when it just has to output the citekeys.

**Comment by omangin** [comment]
With the current implementation `list` reads all files even when it just has to output the citekeys. This can take a long time with a few hundreds of papers. This will be solved by the cache.  In the meanwhile it is also possible to hack list to not build the full repository when only asked for citekeys but this adds complexity to the code that will become useless in the future.

**Comment by 73** [comment]
Since we are talking completion here, the fastest way would be to skip the `list`command and act on a listing of the `bib ` folder with trimmed file endings. One would still need to read the config to determine the location of the folder. Anyway, in my opinion it would make `pubs`a lot more usable if i would not have to type (which on rather long keys means copy & paste) the specific keys.  
--73

**Comment by humm** [comment]
Let's remark that typing part of the citekey *will* autocomplete implicitly if only one correspond to the prefix, and ask for disambiguations with a list of options if not. This is the case for the open command for instance (we may not be the case for all commands). This is *very* close to what you want.

**Comment by 73** [comment]
True. Open fully supports it. Haven't tasted it though. Will test it. Maybe
that will suffice. Will get back to that.
--73

**Comment by 73** [comment]
I think that will suffice. Although it is a bit irritating. Can this stay open with low priority?
--73

**Comment by humm** [comment]
Sure. Current status: need to fix a good technical solution & need to implement #14 first.

**Comment by 73** [comment]
I'll extend that behaviour to all commands. Expect a PR on that.

**Comment by asclepius** [comment]
Anyone still working on autocompletion?  The interface is otherwise great.

**Comment by 73** [comment]
Not at the moment.

**Comment by omangin** [comment]
I finally found some time to get a first working version of autocompletion. It uses [argcomplete](http://argcomplete.readthedocs.io) and hence is supposed to be working both with bash and zsh, although I only tested it on bash so far.

It currently works for the commands, citekeys, and tags. The plugin API still needs to be adapted, and some testing added (although it's not clear how).

Testing, suggestions, and comments are welcome. I opened pull request #61.

Any experience with zsh and argcomplete would also be welcome at that point. It however might require some upstream work ([kislyuk/argcomplete/issues/10](http://github.com/kislyuk/argcomplete/issues/10)).
